### ISSUE TITLE
The Parallel.ForEachAsync propagates OperationCanceledException of unknown origin

### ISSUE THREAD

**Issue by theodorzoulias** [dotnet/runtime#66695]
Hi! I am using the [`Parallel.ForEachAsync`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.parallel.foreachasync) with an asynchronous sequence as source, and I observed occasionally one or more `OperationCanceledException`s to be thrown when an error occurs with the prοcessing of an element. These exceptions are not caused by my own code (the `body` delegate), since I am not watching the `CancellationToken` argument. Below is a minimal demonstration of this behavior, that produces consistently these `OperationCanceledException`s (my actual code produces them only sporadically).

```C#
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

public static class Program
{
    public static void Main()
    {
        static async IAsyncEnumerable<int> GetSequence()
        {
            for (int i = 1; i <= 10; i++) yield return i;
            await Task.Delay(300);
        }

        var options = new ParallelOptions() { MaxDegreeOfParallelism = 10 };
        try
        {
            Parallel.ForEachAsync(GetSequence(), options, async (item, ct) =>
            {
                Console.WriteLine($"Processing #{item}");
                await Task.Delay(100);
                if (item == 5)
                {
                    await Task.Delay(100);
                    throw new Exception();
                }
            }).Wait();
        }
        catch (AggregateException aex)
        {
            foreach (var ex in aex.InnerExceptions)
                Console.WriteLine($"- {ex.GetType().Name}: {ex.Message}");
        }
    }
}
```

Sample output:

```none
Processing #2
Processing #1
Processing #4
Processing #3
Processing #5
Processing #6
Processing #9
Processing #7
Processing #10
Processing #8
- Exception: Exception of type 'System.Exception' was thrown.
- OperationCanceledException: The operation was canceled.
- OperationCanceledException: The operation was canceled.
- OperationCanceledException: The operation was canceled.
- OperationCanceledException: The operation was canceled.
- OperationCanceledException: The operation was canceled.
- OperationCanceledException: The operation was canceled.
- OperationCanceledException: The operation was canceled.
- OperationCanceledException: The operation was canceled.
```

[Live demo](https://dotnetfiddle.net/kT7Rg3).

This is happening only with an `IAsyncEnumerable<T>` as source, not with an `IEnumerable<T>`.

This issue is not the same with [another issue](https://github.com/dotnet/runtime/issues/65963) that I posted a couple of weeks earlier. In that issue the `CancellationToken` argument was observed, and the ensuing `TaskCanceledException`s were generated by my own code. In this issue I have no idea where these `OperationCanceledException`s are coming from.

I would like to ask if this behavior is by design. Thanks!

**Comment by stephentoub** [comment]
> I have no idea where these OperationCanceledExceptions are coming from.

Just print out more of the exception details:
```
---> (Inner Exception #1) System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at System.Threading.SemaphoreSlim.WaitUntilCountOrTimeoutAsync(TaskNode asyncWaiter, Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Parallel.<>c__54`1.<<ForEachAsync>b__54_0>d.MoveNext()<---
```

> I would like to ask if this behavior is by design.
> This issue is not the same with https://github.com/dotnet/runtime/issues/65963 that I posted a couple of weeks earlier

It is effectively the same issue in that we decided not to make any attempt to filter out OperationCanceledExceptions based on their origin, as we can't get it right all the time. `IAsyncEnumerator<T>` is not thread-safe: multiple threads can't all be calling `MoveNextAsync` at the same time, so an async lock is used to gate access to the underlying enumerator.  We pass the CancellationToken into the wait on that lock; then in your example the processing of one item throws an exception, so the token is canceled, and all such waits wake up with a cancellation exception.

Now in this particular case, we are able to isolate the source of the exception and we could ignore it in .NET 7.  But keep in mind there are lots of places that could be observing that token, e.g. it's passed into the enumerator itself and it's passed into every iteration of the loop, so a caller still needs to expect that cancellation exceptions may show up in the aggregate even if they don't request cancellation of the externally-provided token.

**Comment by theodorzoulias** [comment]
@stephentoub OK, thanks for the explanation! From my point of view this behavior of the `Parallel.ForEachAsync` makes the overload that takes an async `source` sequence somewhat quirky. The overload that takes a synchronous sequence feels more robust. Ideally the two overloads should behave the same in case of exceptions IMHO. I would like to be able to switch from sync to async, without having to rethink my exception-handling code!

Would you like me to close this, or keep it open for now until it's decided if it's actionable?

**Comment by stephentoub** [comment]
> Would you like me to close this, or keep it open for now until it's decided if it's actionable?

You can leave it open.  https://github.com/dotnet/runtime/pull/66712 will close it.

> The overload that takes a synchronous sequence feels more robust.

The overload that takes a synchronous enumerable doesn't support canceling waits to access the enumerable.  In that way it's less robust.
